### ISSUE TITLE
Calculate validation loss more frequently

### ISSUE THREAD

**Issue by mateuszpieniak** [allenai/allennlp#4362]
**Is your feature request related to a problem? Please describe.**
I noticed that BERT-like architecture is usually fully fine-tuned, but very shortly (1-2 epochs) with some warm-up (to avoid catastrophic forgetting) - for example in https://arxiv.org/pdf/2004.07180.pdf. I believe that such a resolution is not enough for good validation score estimation i.e. you don't really know the curve's shape. I could lower down learning rate and increase the number of epochs, but it will take a lot of time to train. 

A slightly different example of calculating the validation score more frequently is https://arxiv.org/pdf/1803.09820.pdf, which was implemented in Caffe (no epoch concept). In the paper, the LR test was performed on the whole validation set after N batches.

**Describe the solution you'd like**
Provide an additional argument to `Trainer` that denotes the frequency of validation evaluation.

**Describe alternatives you've considered**
Decrease the learning rate and increase the number of epochs.

**Additional context**
1) It would require some changes for EarlyStopping
2) When I reproduce some papers my validation loss looks like the following
![image](https://user-images.githubusercontent.com/31375424/84766557-50645600-afd1-11ea-847e-a095fd57531c.png)

**Comment by epwalsh** [comment]
I was actually just thinking about this as well. But I think it would be easiest to implement this at the `DataLoader` level. We already *kind of* have a setting for this: `DataLoader.batches_per_epoch`.

But this doesn't quite do what you're asking because if `batches_per_epoch` is less than the total number of batches in the dataset, then you're not guaranteed to see every unique instance:
https://github.com/allenai/allennlp/blob/master/allennlp/data/dataloader.py#L102

That said, I don't think it would too hard to tweak it to ensure that every instance is seen (provided you're running for enough epochs). The `DataLoader` would just have to keep some internal state about where to pick up again next time `iter()` is called.

**Comment by mateuszpieniak** [comment]
@epwalsh Super cool! In my case, I don't need such guarantees. I am not sure though whether `slanted_triangular` scheduler reads the proper number of batches, but I guess I will find out tomorrow.

Btw, out of curiosity, do you think it can be considered a regularization? In theory the bigger the dataset, the better generalization properties. On the other hand sampling data from the training, dataset restricts you from fitting perfectly into your training dataset. This can be considered a regularization technique, but the "effective dataset size" is smaller due to sampling.

**Comment by epwalsh** [comment]
Hmm yea I suppose you could think of it that way. Kind of like dropout, except applied to training instances instead of individual weights.

**Comment by dirkgr** [comment]
`batches_per_epoch` is the right answer here. If that does not guarantee that you'll see all instances, we have to fix it. Is that the case?

**Comment by mateuszpieniak** [comment]
@dirkgr Yes, that's the case. I am satisfied with the current implementation of `batches_per_epoch` for my purpose, but I think that such a guarantee should be more intuitive for a user.

**Comment by epwalsh** [comment]
Ok, sounds like this one can be closed, but I've open up a separate issue to address the behavior with `batches_per_epoch`: https://github.com/allenai/allennlp/issues/4393.
